File: golang/findNonRepeatedChar.go
package main

import "fmt"

func main(){
   fmt.Println(findNonRepeatedChar("abcbae"))
}

func findNonRepeatedChar(str string) string{
    cache := make(map[string]int)
    for _ , v := range str{
        //fmt.Printf("v:%s \n" , string(v))   
        if val, ok := cache[string(v)]; ok{
              //fmt.Printf("v:%s , val: %d\n" , string(v), val)   
              cache[string(v)] = val + 1 
              //fmt.Printf("v:%s , val: %d\n" , string(v), cache[string(v)])   
              continue;
        }
        cache[string(v)] =  1 
     }
     //Printing cache
     //for k, v := range cache{
     //   fmt.Printf("%s: %d\n", k, v)
     //}
     var result = string(' ')
     for _ , v := range str{
         //fmt.Printf("v:%s , val: %d\n" , string(v), cache[string(v)])   
         if( cache[string(v)] == 1){
             result = string(v) 
             break
         }
     } 
     return result
}

File: rust/findNonRepeatedChar.rs
use std::collections::HashMap;

fn main(){

     let s = String::from("abcbae");
     println!("{}", find_non_repeated_char(&s[..]));
}
fn find_non_repeated_char(s :&str) -> char{

    let mut cache = HashMap::<&u8, u32>::new();
    for(_i, item) in s.as_bytes().iter().enumerate()  {
        cache.entry(item).or_insert(0); 
        cache.insert(item, 1 + cache[item]);
    } 
    let mut result =' ';
    for(_i, item) in s.as_bytes().iter().enumerate()  {
        let value = cache[item];
        if value == 1 { 
            result = *item as char; 
            break;
        }
    } 
    result  
}
